File: daily/2026/1292_Maximum_Side_Length_of_a_Square_with_Sum_Less_than_or_Equal_to_Threshold.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    int getrec(vector<vector<int>>& P, int x1, int x2, int y1, int y2){
        return P[x2+1][y2+1] - P[x1][y2+1] - P[x2+1][y1] + P[x1][y1];
    }
    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        int m = mat.size();
        int n = mat[0].size();
        int ans = 0;
        vector<vector<int>> P(m + 1, vector<int>(n + 1, 0));
        for (int i = 0; i < m; i++){
            int acc = 0;
            for (int j = 0; j < n; j++){
                acc += mat[i][j];
                P[i+1][j+1] = P[i][j+1] + acc;
            }
        }
        for (int i = 0; i < m; i++){
            for (int j = 0; j < n; j++){
                for (int e = ans + 1; e <= min(m, n); e++){
                    if (i + e - 1< m && j + e - 1< n && getrec(P, i, i + e - 1, j, j + e - 1) <= threshold) ans = e;
                    else break;
                }
            }
        }
        return ans;
    }
};

int main(){
    vector<vector<int>> mat = {{1,1,3,2,4,3,2},{1,1,3,2,4,3,2},{1,1,3,2,4,3,2}};
    int threshold = 4;
    // vector<vector<int>> mat = {{2,2,2,2,2},{2,2,2,2,2},{2,2,2,2,2},{2,2,2,2,2},{2,2,2,2,2}};
    // int threshold = 1;
    Solution S;
    cout << S.maxSideLength(mat, threshold) << endl;
    return 0;
}

File: daily_python/1292_Maximum_Side_Length_of_a_Square_with_Sum_Less_than_or_Equal_to_Threshold.py
import os
import sys
from typing import List

class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:
        m, n = len(mat), len(mat[0])
        P = [[0] * (n + 1) for _ in range(m + 1)]
        for i in range(m):
            for j in range(n):
                P[i+1][j+1] = P[i][j+1] + P[i+1][j] - P[i][j] + mat[i][j]

        def getRec(x1: int, x2: int, y1: int, y2: int) -> int:
            return P[x2+1][y2+1] - P[x1][y2+1] - P[x2+1][y1] + P[x1][y1]
        
        ans = 0
        for i in range(m):
            for j in range(n):
                for c in range(ans + 1, min(m, n) + 1):
                    if i + c - 1 < m and j + c - 1 < n and getRec(i, i + c - 1, j, j + c - 1) <= threshold:
                        ans = c
                    else:
                        break
        return ans
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            mat = [list(map(int, row.split(','))) for row in lines[0].split('=')[1].strip()[2:-3].split('},{')]
            threshold = int(lines[1].split('=')[1].strip()[:-1])
            dataset.append((mat, threshold))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for mat, threshold in dataset:
        results.append(solution.maxSideLength(mat, threshold))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')